### PR TITLE
Proposal for off-chain message signing

### PR DESCRIPTION
**Pull request by t-nelson**
#### Problem

We don't have a specification for signing off-chain messages, so teams keep rolling their own custom protocol that are various types of dangerous

#### Summary of Changes

Propose a specification for off-chain message signing

cc/ #21366

### CODE REVIEW

**Comment by t-nelson** [comment]
cc/ @fragosti (sry can't add you as reviewer)

File: docs/src/proposals/off-chain-message-signing.md
Comment (terorie) [inline docs/src/proposals/off-chain-message-signing.md:9]: number twister

File: docs/src/proposals/off-chain-message-signing.md
Comment (terorie) [inline docs/src/proposals/off-chain-message-signing.md:53]:
```suggestion
| ID  | Encoding            | Maximum Length | Hardware Wallet Support |
```

File: docs/src/proposals/off-chain-message-signing.md
Comment (terorie) [inline docs/src/proposals/off-chain-message-signing.md:35]: Since the off-chain message signing scheme is targeted at a wide variety of unrelated apps, we might want to force off-chain apps to include their own application identifiers (analogous to ALPN in TLS).

A static signing domain only prevents confusion between on-chain and off-chain messaging. Apps will want separate off-chain signing domains regardless. I propose adding another 16 bytes that the app can choose.

Comment (sdlaver) [reply docs/src/proposals/off-chain-message-signing.md:35]: For format 0 and 1 messages, a fixed field would eat into the (limited) size available for messages. Perhaps a non-normative recommendation to use N bytes at the start of the message to specify app-specific sub-domain?

File: docs/src/proposals/off-chain-message-signing.md
Comment (sdlaver) [inline docs/src/proposals/off-chain-message-signing.md:67]: MAYBE: should there be a section on signing of off-chain messages here?

1. Indicate that Solana messages should be signed with the same Ed25519 signature algorithm as Solana transactions
2. Note that the entire payload (header + message) should be included in the signature calculation
3. Standardize the signature(s) should be appended to the payload

**Comment by buffalu** [comment]
yessss

File: docs/src/proposals/off-chain-message-signing.md
Comment (terorie) [reply docs/src/proposals/off-chain-message-signing.md:35]: Signing domains only work if they are always enforced. Otherwise an app that opts-out can forge a message colliding with all other apps which would not be obvious to the user.
Could you elaborate what limits message length? (hardware wallets I think?)

Comment (sdlaver) [reply docs/src/proposals/off-chain-message-signing.md:35]: Yep, per the message format section, formats 0 and 1 have a limited message length, for support by hardware wallets.

Re: signing domains:
> I propose adding another 16 bytes that the app can choose.

If apps get to choose this, then they can just lie about their domain in any case, right? It's not an area where I have particularly deep expertise, so I will absolutely defer to those more knowledgeable on this, I'm just not clear on how putting this in the header (as a fixed field) vs in the body (as an optional part of the payload, that apps that care about it's presence can check for) differs.

Comment (terorie) [reply docs/src/proposals/off-chain-message-signing.md:35]: > If apps get to choose this, then they can just lie about their domain in any case, right? 

Good point – the hardware wallet could prompt the user to confirm the app ID that's requesting a signature, similar to how on-chain transfers signing requests prompt for the destination address and amount.

**Comment by buffalu** [comment]
somewhat related, but is there a good way to do this that's time based to avoid replay attack? that would fit our use case perfectly

File: docs/src/proposals/off-chain-message-signing.md
Comment (sdlaver) [reply docs/src/proposals/off-chain-message-signing.md:35]: Gotcha - I understand now; a mandatory field presents opportunities for the signer to perform additional validation activities

File: docs/src/proposals/off-chain-message-signing.md
Comment (t-nelson) [reply docs/src/proposals/off-chain-message-signing.md:9]:
```suggestion
[EIP-712](https://eips.ethereum.org/EIPS/eip-712)
```

File: docs/src/proposals/off-chain-message-signing.md
Comment (t-nelson) [reply docs/src/proposals/off-chain-message-signing.md:35]: thoughts on extending the app domain specifier to 32-bytes, so it can support a common pubkey/address?  I know it's more "eating into the message", but this feature is for auth, not writing home to mum.

File: docs/src/proposals/off-chain-message-signing.md
Comment (t-nelson) [reply docs/src/proposals/off-chain-message-signing.md:67]: > 3. Standardize the signature(s) should be appended to the payload

good call. we should specify a container akin to `Transaction`.  may need to support multisig.  i'll noodle on it while i'm ooo in case there's anything else that it should support

File: docs/src/proposals/off-chain-message-signing.md
Comment (jordansexton) [inline docs/src/proposals/off-chain-message-signing.md:43]: What about something generic like `\xffsigned message\xff`?

This format being reusable / multi-chain can help wallets simplify their implementation if the chain they are supporting doesn't already define this.

Comment (t-nelson) [reply docs/src/proposals/off-chain-message-signing.md:43]: that `\xff` is totally solana-specific.  ethereum uses `\x19` for the same purpose.  that is, "shit, what can we use to discriminate this thing that we didn't think about 'til way too late."  I don't think this is a constraint we should force on ourselves

File: docs/src/proposals/off-chain-message-signing.md
Comment (jordansexton) [reply docs/src/proposals/off-chain-message-signing.md:35]: "16 bytes ought to be enough for anybody"

Comment (jordansexton) [reply docs/src/proposals/off-chain-message-signing.md:35]: But wouldn't it need to be 33 bytes, with 0xff prefix?

Comment (t-nelson) [reply docs/src/proposals/off-chain-message-signing.md:35]: > But wouldn't it need to be 33 bytes, with 0xff prefix?

Nah, only the global domain specifier needs that crafted marker ('cause it's the first byte of the payload). App domain specifier can be whatever

File: docs/src/proposals/off-chain-message-signing.md
Comment (terorie) [reply docs/src/proposals/off-chain-message-signing.md:43]: Similar situation with the QUIC node server cert. It still uses the root signing domain (no prefix)

File: docs/src/proposals/off-chain-message-signing.md
Comment (jstarry) [inline docs/src/proposals/off-chain-message-signing.md:38]: Versioned transaction messages always set the first bit and then use the remaining bits of the first byte to determine a version number for the message. This `\xff` value corresponds to version 128 and so we should ensure that the version encoding for transactions spills over into another byte if higher versions are ever used.

Comment (jstarry) [reply docs/src/proposals/off-chain-message-signing.md:38]: Never mind I see that you covered this in the "Runtime Considerations" section below

**Comment by t-nelson** [comment]
> somewhat related, but is there a good way to do this that's time based to avoid replay attack? that would fit our use case perfectly

replay prevention would be out of scope for this proposal.  it requires action on the part of the message issuer whereas we're only trying to define a wire format here

File: docs/src/proposals/off-chain-message-signing.md
Comment (askibin) [reply docs/src/proposals/off-chain-message-signing.md:43]: \x19 was chosen because it is the length of the string "Ethereum Signed Message:\n", which came from Bitcoin's "\x18Bitcoin Signed Message:\n". It turned out later that \x19 was a good choice because RLP encoded transactions can't start with that value. And it was selected for a signed data standard: https://eips.ethereum.org/EIPS/eip-191

**Comment by askibin** [comment]
> > somewhat related, but is there a good way to do this that's time based to avoid replay attack? that would fit our use case perfectly
> 
> replay prevention would be out of scope for this proposal. it requires action on the part of the message issuer whereas we're only trying to define a wire format here

timestamp can be added to the message itself. An app that verifies the signature will receive the timestamp along with the signed message (that includes that timestamp) and check that it is not stale.

File: docs/src/proposals/off-chain-message-signing.md
Comment (terorie) [reply docs/src/proposals/off-chain-message-signing.md:43]: I propose adding a disclaimer such as: "Do not reuse the signing key for any applications except Solana off-chain and transaction signing." 
Multi-chain functionality can be implemented through mutual authentication (i.e. Solana off-chain signature on Other-Chain key and vice versa); or through BIP32-Ed25519 extended public keys/HD wallets.

**Comment by askibin** [comment]
@t-nelson why limit message length for format #2 to 65515? Hypothetically message could be a large document. What if the length is 4 bytes? Also since length is in binary endianess needs to be specified.

File: docs/src/proposals/off-chain-message-signing.md
Comment (t-nelson) [reply docs/src/proposals/off-chain-message-signing.md:43]: > I propose adding a disclaimer such as: "Do not reuse the signing key for any applications except Solana off-chain and transaction signing." Multi-chain functionality can be implemented through mutual authentication (i.e. Solana off-chain signature on Other-Chain key and vice versa); or through BIP32-Ed25519 extended public keys/HD wallets.

not sure that'll fly. afaik the clowns brigading the off-chain message signing for ledger app issue all want to authenticate their nft ownership

**Comment by t-nelson** [comment]
> timestamp can be added to the message itself. An app that verifies the signature will receive the timestamp along with the signed message (that includes that timestamp) and check that it is not stale.

right, but not something we're trying to solve here.  someone can build a message server on top of this proposal, no problem

> @t-nelson why limit message length for format #2 to 65515? Hypothetically message could be a large document. What if the length is 4 bytes?

64k seems like plenty for auth. I alluded to my opinion on large messages [earlier](https://github.com/solana-labs/solana/pull/26915#discussion_r938151622). Any larger content can sign a hash instead of the content itself.  Hypotheticals are great and all (why not 1TB messages!), but when i was doing initial research, i didn't find uses of off-chain signing large messages wrt other popular protocols.

the header is versioned, so we can always extend the max length in the future.

> Also since length is in binary endianess needs to be specified.

Yeah, good point! LE assumption like all the rest of integers in the protocol :slightly_smiling_face:

**Comment by sdlaver** [comment]
@t-nelson anything I can do to help move this proposal along? To avoid turning Mobile Wallet Adapter into a signing oracle, I'd like to incorporate this proposal into the MWA spec as mandatory for signing of non-transaction payloads.

**Comment by YZhenY** [comment]
This makes a ton of sense. Would love to standardise the prefix  @t-nelson so that we could also update Sign in with Solana to it https://github.com/web3auth/sign-in-with-solana

**Comment by fragosti** [comment]
What about https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-122.md

**Comment by CriesofCarrots** [comment]
@t-nelson @askibin , can you please finalize and merge this to document the actual implementation? Or if it should be a simd, move there...

File: docs/src/proposals/off-chain-message-signing.md
Comment (askibin) [inline docs/src/proposals/off-chain-message-signing.md:60]: it should be  `0x20..=0x7e`

Comment (askibin) [reply docs/src/proposals/off-chain-message-signing.md:60]: besides the minor thing I mentioned above implementation matches the proposal, will be good to merge

Comment (t-nelson) [reply docs/src/proposals/off-chain-message-signing.md]: fixed

**Comment by t-nelson** [comment]
updated to include
* all outstanding review comments
* all incompletely specified details
  * add application domain specifier to preamble
  * add signer list to preamble
  * add specification for signing and verification
  * add envelope format for attaching signatures to messages

pending review, i believe this document is now _finally_ complete.  we should eventually port it over to SIMD along with the other proposals.

cc/ @askibin @fragosti @terorie @jordansexton @sdlaver

**Comment by askibin** [comment]
when a server will generate a message to sign, I imagine it will need to add some volatile data to prevent replay attacks. E.g., a timestamp or a nonce. The server can include them as a part of the message, but this way they will be displayed to the user without having practical meaning for them. It might make sense to have a separate field in the Preamble for it. Also, the server can impose an expiration time, it might be useful if a client can see when the message is no longer valid.

**Comment by t-nelson** [comment]
> when a server will generate a message to sign, I imagine it will need to add some volatile data to prevent replay attacks. E.g., a timestamp or a nonce. The server can include them as a part of the message, but this way they will be displayed to the user without having practical meaning for them. It might make sense to have a separate field in the Preamble for it. Also, the server can impose an expiration time, it might be useful if a client can see when the message is no longer valid.

i'm still of the opinion from this comment

https://github.com/solana-labs/solana/pull/26915#issuecomment-1224933935

i'm envisioning multiple message formats for various protocols being built atop what's proposed here. i'd rather limit this proposal's requirements to the minimum required for a concise, extensible, secure, signed message. folks can spec out extensions if the use case has sufficient demand as to justify that effort. if it ends up that every extension uses some common field(s), we have a version byte for that :wink:

**Comment by t-nelson** [comment]
updated to impose non-zero requirement on `signers length` and `message length` (already required by the implementation

**Comment by t-nelson** [comment]
no complaints in two weeks.  who has approval for me?

File: docs/src/proposals/off-chain-message-signing.md
Comment (askibin) [inline docs/src/proposals/off-chain-message-signing.md:125]: shall we get rid of this paragraph given the #29807

Comment (t-nelson) [reply docs/src/proposals/off-chain-message-signing.md:125]: i think it's ok to leave it as a reference to what needed to be changed to accommodate offchain message signing. i'll add a link to your PR implementing it though.  good catch!

Comment (t-nelson) [reply docs/src/proposals/off-chain-message-signing.md:125]: https://github.com/solana-labs/solana/pull/26915/commits/c0857968ebf5a5be2dc24953389fccbe4e93217a#diff-e99c27ae1763d88fc8e45e699b94c3a9b4e8ccdab3a4a425d1d3db5012e70b7aR136-R138

**Review by askibin — APPROVED** [review]

**Pull request merged**
